core/commands/get.go

```diff
@@ -0,0 +1,426 @@
+		compress = (compress && (compressionLevel > 0 || !found)) || compressionLevel > 0
+
+		if archive, _, _ := req.Option("archive").Bool(); archive {
+			if !strings.HasSuffix(outPath, ".tar") {
+				outPath += ".tar"
+			}
+			if compress {
+				outPath += ".gz"
+			}
+			fmt.Printf("Saving archive to %s\n", outPath)
+
+			file, err := os.Create(outPath)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+			defer file.Close()
+
+			bar := pb.New(0).SetUnits(pb.U_BYTES)
+			bar.Output = os.Stderr
+			pbReader := bar.NewProxyReader(reader)
+			bar.Start()
+			defer bar.Finish()
+
+			_, err = io.Copy(file, pbReader)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			return
+		}
+
+		fmt.Printf("Saving file(s) to %s\n", outPath)
+
+		// TODO: get total length of files
+		bar := pb.New(0).SetUnits(pb.U_BYTES)
+		bar.Output = os.Stderr
+
+		preexisting := true
+		pathIsDir := false
+		if stat, err := os.Stat(outPath); err != nil && os.IsNotExist(err) {
+			preexisting = false
+		} else if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		} else if stat.IsDir() {
+			pathIsDir = true
+		}
+
+		var tarReader *tar.Reader
+		if compress {
+			gzipReader, err := gzip.NewReader(reader)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+			defer gzipReader.Close()
+			pbReader := bar.NewProxyReader(gzipReader)
+			tarReader = tar.NewReader(pbReader)
+		} else {
+			pbReader := bar.NewProxyReader(reader)
+			tarReader = tar.NewReader(pbReader)
+		}
+
+		bar.Start()
+		defer bar.Finish()
+
+		for i := 0; ; i++ {
+			header, err := tarReader.Next()
+			if err != nil && err != io.EOF {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+			if header == nil || err == io.EOF {
+				break
+			}
+
+			if header.Typeflag == tar.TypeDir {
+				pathElements := strings.Split(header.Name, "/")
+				if !preexisting {
+					pathElements = pathElements[1:]
+				}
+				path := fp.Join(pathElements...)
+				path = fp.Join(outPath, path)
+				if i == 0 {
+					outPath = path
+				}
+
+				err = os.MkdirAll(path, 0755)
+				if err != nil {
+					res.SetError(err, cmds.ErrNormal)
+					return
+				}
+				continue
+			}
+
+			var path string
+			if i == 0 {
+				if preexisting {
+					if !pathIsDir {
+						res.SetError(os.ErrExist, cmds.ErrNormal)
+						return
+					}
+					path = fp.Join(outPath, header.Name)
+				} else {
+					path = outPath
+				}
+			} else {
+				pathElements := strings.Split(header.Name, "/")[1:]
+				path = fp.Join(pathElements...)
+				path = fp.Join(outPath, path)
+			}
+
+			file, err := os.Create(path)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			_, err = io.Copy(file, tarReader)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			err = file.Close()
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+		}
```

Comment:
a ton of this can be broken out into util funcs in small pkgs inside `thirdparty/`. For example:

```go

type TarExpander struct {
  Path string
}

func (te *TarExpander) Expand(r tar.Reader) error {}
func (te *TarExpander) ExpandDir(h tar.Header) error {}
func (te *TarExpander) ExpandFile(h tar.Header, r tar.Reader) error {}
```